File: runtime/core/scheduling.py
"""Dependency and write-ownership checks shared by all orchestration modes."""

from __future__ import annotations

from pathlib import PurePosixPath
from typing import Iterable

from .contracts import ContractError, WorkItem


def normalize_scope(scope: str) -> PurePosixPath:
    value = scope.strip().replace("\\", "/").strip("/")
    if not value:
        raise ContractError("ownership scopes must not be empty")
    path = PurePosixPath(value.casefold())
    if ".." in path.parts:
        raise ContractError(f"ownership scope escapes its root: {scope}")
    return path


def scopes_overlap(left: str, right: str) -> bool:
    a = normalize_scope(left)
    b = normalize_scope(right)
    return a == b or a in b.parents or b in a.parents


def validate_work_graph(items: Iterable[WorkItem]) -> dict[str, WorkItem]:
    by_id: dict[str, WorkItem] = {}
    for item in items:
        if item.id in by_id:
            raise ContractError(f"duplicate work item id: {item.id}")
        by_id[item.id] = item
    if not by_id:
        raise ContractError("managed run requires at least one work item")
    for item in by_id.values():
        if item.id in item.depends_on:
            raise ContractError(f"{item.id}: self dependency")
        for dependency in item.depends_on:
            if dependency not in by_id:
                raise ContractError(f"{item.id}: unknown dependency {dependency}")
        for scope in item.ownership:
            normalize_scope(scope)

    visiting: set[str] = set()
    visited: set[str] = set()

    def visit(item_id: str) -> None:
        if item_id in visiting:
            raise ContractError(f"dependency cycle contains {item_id}")
        if item_id in visited:
            return
        visiting.add(item_id)
        for dependency in by_id[item_id].depends_on:
            visit(dependency)
        visiting.remove(item_id)
        visited.add(item_id)

    for item_id in by_id:
        visit(item_id)
    return by_id


def validate_parallel_wave(items: Iterable[WorkItem]) -> None:
    writers = [item for item in items if item.mode == "write"]
    for index, left in enumerate(writers):
        for right in writers[index + 1 :]:
            for left_scope in left.ownership:
                for right_scope in right.ownership:
                    if scopes_overlap(left_scope, right_scope):
                        raise ContractError(f"write ownership overlaps between {left.id} and {right.id}")

File: src/runtime/core/scheduling.ts
import { ContractError, type WorkItem } from "./contracts.js";

export function normalizeScope(scope: string): string[] {
  const normalized = scope.trim().replaceAll("\\", "/").replace(/^\/+|\/+$/gu, "").toLocaleLowerCase("en-US");
  if (normalized.length === 0) {
    throw new ContractError("ownership scopes must not be empty");
  }
  const parts = normalized.split("/").filter((part) => part.length > 0 && part !== ".");
  if (parts.includes("..")) {
    throw new ContractError(`ownership scope escapes its root: ${scope}`);
  }
  return parts;
}

export function scopesOverlap(left: string, right: string): boolean {
  const a = normalizeScope(left);
  const b = normalizeScope(right);
  const prefixLength = Math.min(a.length, b.length);
  return a.slice(0, prefixLength).every((part, index) => part === b[index]);
}

export function validateWorkGraph(items: Iterable<WorkItem>): Record<string, WorkItem> {
  const byId: Record<string, WorkItem> = {};
  for (const item of items) {
    if (item.id in byId) {
      throw new ContractError(`duplicate work item id: ${item.id}`);
    }
    byId[item.id] = item;
  }
  if (Object.keys(byId).length === 0) {
    throw new ContractError("managed run requires at least one work item");
  }
  for (const item of Object.values(byId)) {
    if (item.depends_on.includes(item.id)) {
      throw new ContractError(`${item.id}: self dependency`);
    }
    for (const dependency of item.depends_on) {
      if (!(dependency in byId)) {
        throw new ContractError(`${item.id}: unknown dependency ${dependency}`);
      }
    }
    item.ownership.forEach(normalizeScope);
  }

  const visiting = new Set<string>();
  const visited = new Set<string>();
  const visit = (id: string): void => {
    if (visiting.has(id)) throw new ContractError(`dependency cycle contains ${id}`);
    if (visited.has(id)) return;
    visiting.add(id);
    for (const dependency of byId[id]?.depends_on ?? []) visit(dependency);
    visiting.delete(id);
    visited.add(id);
  };
  Object.keys(byId).forEach(visit);
  return byId;
}

export function validateParallelWave(items: Iterable<WorkItem>): void {
  const writers = Array.from(items).filter((item) => item.mode === "write");
  for (let leftIndex = 0; leftIndex < writers.length; leftIndex += 1) {
    const left = writers[leftIndex];
    if (left === undefined) continue;
    for (const right of writers.slice(leftIndex + 1)) {
      if (left.ownership.some((a) => right.ownership.some((b) => scopesOverlap(a, b)))) {
        throw new ContractError(`write ownership overlaps between ${left.id} and ${right.id}`);
      }
    }
  }
}
